From 2f8b8c266408490f8582341952e0ddb836e6dfb8 Mon Sep 17 00:00:00 2001
From: kabirz <jxwazxzhp@126.com>
Date: Mon, 11 Aug 2025 12:44:21 +0800
Subject: [PATCH 1/2] serail: databit 7 stopbit 1 parity even

Signed-off-by: kabirz <jxwazxzhp@126.com>
---
 drivers/serial/uart_stm32.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/serial/uart_stm32.c b/drivers/serial/uart_stm32.c
index 0177bcb029f..47b2e09f240 100644
--- a/drivers/serial/uart_stm32.c
+++ b/drivers/serial/uart_stm32.c
@@ -413,7 +413,7 @@ static inline uint32_t uart_stm32_cfg2ll_databits(enum uart_config_data_bits db,
 	case UART_CFG_DATA_BITS_8:
 	default:
 #ifdef LL_USART_DATAWIDTH_9B
-		if (p != UART_CFG_PARITY_NONE) {
+		if (p != UART_CFG_PARITY_NONE && db != UART_CFG_DATA_BITS_7) {
 			return LL_USART_DATAWIDTH_9B;
 		}
 #endif
@@ -907,6 +907,9 @@ static void fifo_read_with_u8(USART_TypeDef *usart, void *rx_data, const int off
 	uint8_t *data = (uint8_t *)rx_data;
 
 	data[offset] = LL_USART_ReceiveData8(usart);
+	if (LL_USART_GetParity(usart) != LL_USART_PARITY_NONE &&
+		LL_USART_GetDataWidth(usart) == LL_USART_DATAWIDTH_8B)
+		data[offset] &= 0x7f;
 }
 
 static int uart_stm32_fifo_read(const struct device *dev, uint8_t *rx_data, const int size)
-- 
2.43.0

